src/krb5/_creds_mit.pyx

```diff
@@ -120,3 +151,55 @@ EtypeInfo = collections.namedtuple('EtypeInfo', [
     'salt',
     's2kparams',
 ])
+
+def serialize_creds(Context context not None, Creds self not None) -> bytes:
+    cdef krb5_error_code err = 0
+    cdef krb5_data *data = NULL
+    cdef size_t length
+    cdef char *value
+
+    try:
+        err = krb5_marshal_credentials(context.raw, &self.raw, &data)
+
+        if err:
+            raise Krb5Error(context, err)
+
+        pykrb5_get_krb5_data(data, &length, &value)
+
+        if length == 0:
+            data_bytes = b""
+        else:
+            data_bytes = value[:length]
+
+        return data_bytes
+
+    finally:
+        if NULL != data:
+            pykrb5_free_data(context.raw, data)
+
+def unserialize_creds(Context context not None, const unsigned char[:] data not None) -> Creds:
+    cdef krb5_error_code err = 0
+    cdef krb5_creds *creds_raw = NULL
+    cdef krb5_data data_raw
+
+    creds = Creds(context)
+
+    if len(data) == 0:
+        pykrb5_set_krb5_data(&data_raw, 0, "")
+    else:
+        pykrb5_set_krb5_data(&data_raw, len(data), <char *>&data[0])
+
+    err = krb5_unmarshal_credentials(context.raw, &data_raw, &creds_raw)
+
+    if creds_raw:
+        # creds_raw was calloc'ed for krb5_creds structure
```

Comment:
I'm somewhat weary about using `free` here explicitly. The docs for [krb5_unmarshal_credentials](https://web.mit.edu/kerberos/krb5-devel/doc/appdev/refs/api/krb5_unmarshal_credentials.html) state that

> Use krb5_free_creds() to free creds_out when it is no longer needed.

I can see the actual implementation essentially calls `krb5_free_creds_content` and then calls `free`. https://github.com/krb5/krb5/blob/354f176ba6d6cc544e1c15712a13f9c006ca605d/src/lib/krb5/krb/kfree.c#L213-L219

While this replicates that same logic that's more of an implementation detail and not a public API. We should try to avoid freeing data using `free` if we don't control how it was allocated.

Instead I'm wondering if we should add another `cdef` field on `Creds` that can be used to control how the struct is deallocated to work with this new scenario. For example (the field it actually defined in the pxd file, the comment is more for documentation)

```python
cdef class Creds:
    # cdef Context ctx
    # cdef krb5_creds raw
    # cdef int needs_free
    # cdef int free_struct

    def __dealloc__(Creds self):
        if self.needs_free:
            if self.free_struct:
                krb5_free_creds(self.ctx.raw, &self.raw)
            else:
                krb5_free_cred_contents(self.ctx.raw, &self.raw)
            self.needs_free = 0
```

The `krb5_free_creds` would also need to be defined in the file but seems like Heimdal has it so it shouldn't be a problem.

The other benefit of this is we don't need to copy the data we can just do something like

```python
creds = Creds(context)

...

err = krb5_unmarshal_credentials(context.raw, &data_raw, &creds.raw)
if err:
    raise Krb5Error(context, err)

creds.free_struct = 1
creds.needs_free = 1

return creds
```

Do you see any problem with this approach?

Reply:
I agree I stepped onto thin ice here... The point is a _pointer-to-pointer_ used by `krb5_unmarshal_credentials()`, it unconditioanaly allocates the structure using its `k5alloc()` wrapper and writes the allocated pointer to `creds_out`. There is just no way to not allocate some heap memory, so it has to be free'd. 

https://github.com/krb5/krb5/blob/354f176ba6d6cc544e1c15712a13f9c006ca605d/src/lib/krb5/ccache/ccmarshal.c#L557 

There are might be two ways to avoid shortcuts like mine, right after `krb5_unmarshal_credentials()`
1) Deep copy the new creds to our `self.raw` and deallocate everything just returned with `krb5_free_creds() `  as original API prescribes. Our existing `raw` will contain a deep copy to be free'd with `krb5_free_cred_contents()` in `__deallocate__`.
2) Shallow copy the new creds to our `self.raw`, save the alloc'ed pointer as `_raw_ptr` and indeed call  `krb5_free_creds` instead of `krb5_free_cred_contents` in `__deallocate__`.  Our existing `raw` will remain the shallow copy to keep other things working. 

The problem with (1) is that deep copy function `krb5_copy_creds()` behaves the same as `krb5_unmarshal_credentials()` and allocates the new structure. So I came to conclusion that additional pointer has to be stored in object, and touching it will obviously crash the whole process. I decided to do another risky shortcut but I can of course change.

It is up to you to decide, are you ok with having additional `_raw_ptr` member in `Creds` class?
Or any better ideas?

Reply:
Ah I see, sorry I didn't see it was a `krb5_creds**` argument. I thought it was just `krb5_creds*`, now the API makes more sense.